modules/oracle/marketObserve.go

```diff
@@ -0,0 +1,204 @@
+package oracle
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"math"
+	"time"
+	"vsc-node/modules/oracle/p2p"
+
+	"github.com/go-playground/validator/v10"
+)
+
+var (
+	v = validator.New(validator.WithRequiredStructEnabled())
+
+	// to be processed by the block producer, calculating median price
+	// and creating new block
+	broadcastPriceBuf = make([]p2p.AveragePricePoint, 0, 256)
+
+	// to be signed by the witness
+	newBlockBuf = make([]p2p.VSCBlock, 0, 256)
+)
+
+func (o *Oracle) marketObserve() {
+	pricePollTicker := time.NewTicker(priceOraclePollInterval)
+
+	for {
+		select {
+		case <-o.ctx.Done():
+			return
+
+		case <-pricePollTicker.C:
+			for _, api := range o.priceOracle.PriceAPIs {
+				go api.QueryMarketPrice(watchSymbols, o.observePriceChan)
+			}
+
+		case broadcastSignal := <-o.broadcastPriceSignal:
+			o.handleBroadcastSignal(broadcastSignal)
+			broadcastPriceBuf = broadcastPriceBuf[:0]
+			newBlockBuf = newBlockBuf[:0]
+			o.priceOracle.ResetPriceCache()
+
+		case newBlock := <-o.broadcastPriceBlockChan:
+			// TODO: move this channel to witness processing
+			newBlockBuf = append(newBlockBuf, newBlock)
+
+		case avgPricePoints := <-o.broadcastPriceChan:
+			now := time.Now().UTC().Unix()
+			for i := range avgPricePoints {
+				avgPricePoints[i].UnixTimeStamp = now
+			}
+			broadcastPriceBuf = append(broadcastPriceBuf, avgPricePoints...)
+
+		case msg := <-o.msgChan:
```

Comment:
Unnecessary use of channels

related to note 1.3

Reply:
Since it looks like you are trying to do P2P communication you can just directly send to the P2P APIs. There is no reason to put in a channel in-between the code wanting to send a message and the p2p service

Reply:
If you want to create some middleware between the two that is OK, but that is completely different than forcing it through a channel